File: CppPractice/Project1/Project1/Leetcode206.cpp
#include <string>

#include <vector>

#include <iostream>

#include <stdlib.h>

#include <time.h>

#include <sstream>

#include <algorithm>

#include <map>

#include <bitset>

using namespace std;


struct ListNode {
    int val;
    ListNode* next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode* next) : val(x), next(next) {}
};


//class Solution {
//public:
//    ListNode* reverseList(ListNode* head) {
//        ListNode* ans = new ListNode();
//        ListNode* cur1 = head1;
//        ListNode* cur2 = head2;
//        int carry = 0;
//        while (cur1 or cur2 or carry==1)
//        {
//           
//            ans->val = (carry + cur1->val + cur2->val)%10;
//            carry = (carry + cur1->val + cur2->val) / 10;
//            if (cur1)
//            {
//                cur1 = cur1->next;
//
//            }
//            if (cur2)
//            {
//                cur2 = cur2->next;
//            }
//            ans->next = new ListNode();
//            ans = ans->next;
//        }
//
//        return ans;
//
//
//    }
//};


class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        
       ListNode* ans = new ListNode();
       ListNode* cur = head;
       vector<int> val_list;
       if (!head)
       {
           return head;

       }
       while (cur!=NULL)
       {
           val_list.push_back(cur->val);
           cur = cur->next;
       }
       reverse(val_list.begin(), val_list.end());
       cur = ans;
       for (int i = 0; i < val_list.size(); i++)
       {
           cur->next = new ListNode(val_list[i]);
           cur = cur->next;

       }
       return ans->next;

    }
};

File: Pythonleetcode/Leetcode206.py
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution(object):
    def reverseList(self, head1,head2):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        ans=ListNode()
        p1=head1
        p2=head2
        carry=0
        while p1 or p2 or carry==1:
            ans.next = ListNode()
            ans.val = carry
            if p1:
                ans.val+=p1.val
            if p2:
                ans.val+=p2.val
            carry=ans.val//10
            ans.val=ans.val%10
            print(ans.val)
            if p1:
                p1=p1.next
            if p2:
                p2=p2.next
        return ans

head1=ListNode(1)
head1.next=ListNode(1)
head1.next.next=ListNode(2)
head1.next.next.next=ListNode(3)

head2=ListNode(5)
head2.next=ListNode(1)
head2.next.next=ListNode(2)
head2.next.next.next=ListNode(9)
head2.next.next.next.next=ListNode(3)

print(Solution().reverseList(head1,head2))









# 修改val方法
# class Solution(object):
#     def reverseList(self, head):
#         """
#         :type head: ListNode
#         :rtype: ListNode
#         """
#         pointer=head
#         val_list=[]
#         while pointer!=None:
#             val_list.append(pointer.val)
#             pointer=pointer.next
#         val_list=val_list[::-1]
#         pointer=head
#         p_index=0
#         while pointer!=None:
#             pointer.val=val_list[p_index]
#             pointer=pointer.next
#             p_index+=1
#         return head


# 三指针法
# class Solution(object):
#     def reverseList(self, head):
#         """
#         :type head: ListNode
#         :rtype: ListNode
#         """
#         if not head or not head.next:
#             return head
#         if not head.next.next:
#             first=head
#             second=head.next
#             first.next=None
#             second.next=first
#             head=second
#             return head
#         first=head
#         second=head.next
#         third=head.next.next
#         first.next=None
#         while third!=None:
#             second.next=first
#             first=second
#             second=third
#             third=third.next
#             # print(first.val,second.val,third.val)
#         second.next=first
#         head=second
#         return head












